### ISSUE TITLE
nn.MultiheadAttention breaks for mask_type=2 when fast path is enabled (e.g. with batched attn_mask)

### ISSUE THREAD

**Issue by lukasschmit** [pytorch/pytorch#97409]
### 🐛 Describe the bug

**TLDR:** When `nn.MultiheadAttention` is used with a batched `attn_mask` which [should be](https://pytorch.org/docs/stable/generated/torch.nn.MultiheadAttention.html) shape (N*H, L, S) (where S=L for self-attn) **_and_** fast path is enabled it crashes.

_It works as expected when fast path is **not** enabled_

### Minimal example to reproduce:

```python
import torch
import torch.nn as nn


with torch.no_grad():
    B = 8
    L = 100
    D = 512
    H = 8

    mha = nn.MultiheadAttention(D, H, batch_first=True)

    X = torch.randn(B, L, D)
    M = torch.randn(B * H, L, L) > 0

    mha.train()  # disable fast path
    out, _ = mha(X, X, X, attn_mask=M, need_weights=False)  # works
    mha.eval()  # enable fast path
    out, _ = mha(X, X, X, attn_mask=M, need_weights=False)  # crashes
```

Gives:
```
RuntimeError: For mask_type == 2 mask shape should match input shape
```

I did a little debugging in the torch code around [here](https://github.com/pytorch/pytorch/blob/fe0afc58529bdf148bfda580a60e3d778b5b8f1d/torch/nn/modules/activation.py#L1176) and it seems that when `self.merge_masks` is called the `mask_type` returned is 0. However, when `torch._native_multi_head_attention` is called with `mask_type=0` but a (N*H, L, S) `attn_mask`, it seems to ignore the `mask_type` flag (see error above).

I'm on pytorch 2.0 (conda) on M2 Max macOS 13.2.1.

### Versions

Collecting environment information...
PyTorch version: 2.0.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.2.1 (arm64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.16 (main, Jan 11 2023, 10:02:19)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-13.2.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M2 Max

Versions of relevant libraries:
[pip3] lion-pytorch==0.0.4
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==2.0.0
[pip3] pytorch-wavelets==1.3.0
[pip3] torch==2.0.0
[pip3] torch-yin==0.1.3
[pip3] torchaudio==2.0.0
[pip3] torchmetrics==0.11.1
[pip3] xitorch==0.4.0.dev0+3327cc0
[conda] lion-pytorch              0.0.4                    pypi_0    pypi
[conda] numpy                     1.23.5           py39h1398885_0  
[conda] numpy-base                1.23.5           py39h90707a3_0  
[conda] pytorch                   2.0.0                   py3.9_0    pytorch
[conda] pytorch-lightning         2.0.0                    pypi_0    pypi
[conda] pytorch-wavelets          1.3.0                    pypi_0    pypi
[conda] torch-yin                 0.1.3                    pypi_0    pypi
[conda] torchaudio                2.0.0                  py39_cpu    pytorch
[conda] torchmetrics              0.11.1                   pypi_0    pypi
[conda] xitorch                   0.4.0.dev0+3327cc0          pypi_0    pypi

cc @ezyang @gchanan @zou3519 @jbschlosser @bhosmer @cpuhrsch @erichan1

**Comment by angusturner** [comment]
Not exactly the same, but I am also having issues with the fast path since upgrading to 2.0:
#97128

**Comment by cpuhrsch** [comment]
cc @mikekgfb

**Comment by cpuhrsch** [comment]
Marking this for high priority due to traffic to make sure it gets discussed during triage review. cc @jisaacso @drisspg @jbschlosser

**Comment by cpuhrsch** [comment]
@erichan1 - Once a fix is out can you mark it for cherry pick as well? cc @atalman

**Comment by jbschlosser** [comment]
Posted guidelines for a proper fix [here](https://github.com/pytorch/pytorch/pull/98375#issuecomment-1499504721) on the open PR.
